[material-ui] Add theme.focusVisible opt-in keyboard focus ring (#48743)
mui-org/material-ui: docs/data/material/customization/focus-visible/focus-visible.md

> # Focus visible

Reviewer: there are some components throughout this page that have their value controlled, while others don't. is there a reason for going one way or another, or can we be consistent? for example the box shadow tabs, or the toggle button group in the full demo.

Author: Thanks, fixed.

> <p class="description">Configure a theme option to apply an outline keyboard focus across Material UI components.</p>

Reviewer 2: Simplification and correctness suggestion

```suggestion
<p class="description">Apply an outline or inset keyboard focus across Material UI components.</p>
```

Author: fixed in 84bec3c2ac.

Right about the inset, the ring is not always an outline.

One change from the suggestion: kept the non-breaking space in Material UI. A normal space fails the vale `MUI.MuiBrandName` rule.

> Starting from v9.x, Material UI provides built-in support for visual keyboard focus indicator through CSS. The demos on this page opt out of the ripple to show only the focus visible indicator.

Reviewer 2: ```suggestion
Starting from v9.4, Material UI provides built-in support for visual keyboard focus indicator through CSS. The demos on this page opt out of the ripple to show only the focus visible indicator.
```

Author: fixed in 84bec3c2ac — pinned to v9.4.

:::

## Full focus visible demo

The complete set of components that render the focus indicator when `focusVisible` is enabled. Use the keyboard (<kbd>Tab</kbd> and arrow keys) to move focus and reveal the ring.

{{"demo": "FullFocusVisibleDemo.js"}}

## Caveats

### Checkbox and Radio custom icons must be SVG

The Checkbox and Radio attach the focus indicator to the first `<svg>` element inside the component. When customizing them with the `icon` and `checkedIcon` props, the custom icon must render an `<svg>` element — icons rendered as other elements, such as font icons or `<img>`, do not receive the focus indicator.

The indicator hugs whatever box the svg renders at, so smaller replacement icons get a proportionally tighter ring with no extra tuning.

:::success
[`SvgIcon`](/material-ui/icons/#svgicon) is recommended to wrap custom svgs to get consistent styles.
:::

{{"demo": "FocusVisibleCustomIcons.js"}}

### Component focus-visible styles are replaced by the theme

> Some components indicate keyboard focus with a translucent background or overlay by default — the Chip, MenuItem, ListItemButton, AccordionSummary, PaginationItem, CardActionArea, Autocomplete options, and the Slider thumb. When `focusVisible` is enabled, these component focus-visible styles are removed so that the theme's indicator is the only one, consistent across all components — hover, selected, and active styles are unchanged.

Reviewer 2: If Chip, MenuItem and others are components, then we should wrap them in \`\`.
Or we need to add spaces (i.e. "Menu Item", "Accordion Summary").

Author: fixed in 84bec3c2ac — backticked them, same as the rest of the page.
